compiler/vm/vmdef.nim

```diff
@@ -121,17 +319,61 @@ type
     slots*: ptr UncheckedArray[TFullReg]
     currentException*: PNode
     currentLineInfo*: TLineInfo
+
+    # XXX: These are only here as a temporary measure until I continue 
+    # to work on callback handling
+    typeCache*: ptr TypeInfoCache
+    mem*: ptr VmMemoryManager
+
   VmCallback* = proc (args: VmArgs) {.closure.}
 
+  SlotStorage* = object ## A fixed size container for storing registers 
+                        ## that also allows for shallow copies
+    slots: ptr UncheckedArray[TFullReg]
+    length: int
+    owner: bool ## Indicates whether an instance of `SlotStorage` is the owner of the backing memory
```

Comment:
Correct me if I'm wrong, but during execution, slots are basically inherited from the previous frame with a few extras, and there are no disjoint frames, as in only `main->foo->bar` exists and not `main->foo` and `main->bar` at the same time?

If that's the case, I think a better way is to have `TCtx` holds all available slots as a `seq`, then each frame holds the starting offset as well and the length, or use a `Slice[T]` to hold the beginning and the end offsets. That way, you never deal with pointers and avoid having to write extra code to manage that memory.

Reply:
Unfortunately, due to how raising exceptions works, disjoint frames exist (I left exception handling untouched for now). There's also the issue that the address of slots is taken, which requires the backing storage to stay at the same location (could of course also store an index into the slot list instead). I already know how to fix this (get rid of `rkRegAddr`, which I'd consider a hack), but I'd need to do larger adjustments to `vmgen` for that, which I won't do as part of this PR.

If the issues mentioned above are fixed, what you proposed would work, but it might still make sense to let the frame own it's slots (e.g. for closure iterators or resume-able functions in general). Another reason why I used a custom container for slot storage was, to have a shallow copy-able sequence type that works with both `refc` and `orc` (as `seq` doesn't)

Reply:
> it might still make sense to let the frame own it's slots (e.g. for closure iterators or resume-able functions in general).

Resumable functions already have a runner function and an environment passed as a parameter, so I don't think they need to hold their own slots while they are not running.

> Another reason why I used a custom container for slot storage was, to have a shallow copy-able sequence type that works with both `refc` and `orc` (as `seq` doesn't)

If you use one central register storage, then shallow copy should stop being relevant.

Reply:
> Resumable functions already have a runner function and an environment passed as a parameter, so I don't think they need to hold their own slots while they are not running.

If we use them in lowered form, then yes. But I thought that it'd make sense to implement them directly in the VM for efficiency (rather easy, since we already manage stack frames)

Reply:
Looking very far down the road storing the offset and limit makes frames a fixed size allocation. So eventually that's where we should probably head, eventually.

Doesn't need to happen now, especially given the state of vmgen.